README.rst
*****************************
Python Free Energy Estimators
*****************************

.. image:: https://travis-ci.org/markovmodel/pyfeat.svg?branch=devel
    :target: https://travis-ci.org/markovmodel/pyfeat
.. image:: https://coveralls.io/repos/markovmodel/pyfeat/badge.svg?branch=devel
   :target: https://coveralls.io/r/markovmodel/pyfeat?branch=devel
.. image:: https://badge.fury.io/py/pyfeat.svg
   :target: https://pypi.python.org/pypi/pyfeat

The python free energy analysis toolkit has been reactivated from pyfeat https://github.com/markovmodel/pyfeat

**This code is under very heavy development!**

**Note**:

Currently the following algorithms are featured:

- WHAM
- DTRAM
- XTRAM

In future releases the list of available free energy algorithms will be extended. Watch this space!  


Installation
============

TBD


Authors
=======

Current developers:
-------------------
- Antonia Mey <antonia.mey@ed.ac.uk>


Original developers:
--------------------
- Antonia Mey\ :superscript:`*` <antonia.mey@fu-berlin.de>
- Christoph Wehmeyer\ :superscript:`*` <christoph.wehmeyer@fu-berlin.de>
- Fabian Paul
- Hao Wu
- Frank Noé

``*``) equal contribution

Original installation:
----------------------
Valid for up to version 0.3.1 with Python 2.7 dependency

With pip from PyPI ::

   # you might have to install these dependencies manually
   pip install cython
   pip install numpy
   pip install pytram

   # install pyfeat
   pip install pyfeat

References
==========

* **dTRAM**: *Statistically optimal analysis of state-discretized trajectory data from multiple thermodynamic states*, Hao Wu, Antonia S.J.S. Mey, Edina Rosta, and Frank Noé, **J. Chem. Phys.** 141, 214106 (2014). 

    Download: <http://scitation.aip.org/content/aip/journal/jcp/141/21/10.1063/1.4902240>

* **xTRAM**: *Estimating Equilibrium Expectations from Time-Correlated Simulation Data at Multiple Thermodynamic States*, Antonia S.J.S. Mey, Hao Wu, and Frank Noé, **Phys. Rev. X** 4, 041018 (2014). 

    Download: <http://journals.aps.org/prx/pdf/10.1103/PhysRevX.4.041018>

* **WHAM**:  *The weighted histogram analysis method for free-energy calculations on biomolecules. I. The method*, Shankar Kumar, John M. Rosenberg, Djamal Bouzida, Robert H. Swendsen and Peter A. Kollman, **J. Comput. Chem.** 13, 1011–1021 (1992)

    Download: <http://onlinelibrary.wiley.com/doi/10.1002/jcc.540130812/abstract;jsessionid=096E6ED4241821B755F9424DCE203430.f03t02>

Copyright notice
================

Copyright (c) 2020, Mey Research, University of Edinburgh, Edinburgh, EH9 3FJ, UK
Copyright (c) 2014, Computational Molecular Biology Group, FU Berlin, 14195 Berlin, Germany.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions
are met:

1. Redistributions of source code must retain the above copyright notice,
this list of conditions and the following disclaimer.

2. Redistributions in binary form must reproduce the above copyright
notice, this list of conditions and the following disclaimer in the
documentation and/or other materials provided with the distribution.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.




CHANGELOG
*********
CHANGELOG
*********


0.1.x
=====

Implemented WHAM and MBAR (internal Alpha version - not released)


0.2.0
=====

WHAM + MBAR functional (Beta version)

0.2.1
-----

Switching dependencies to numpy>=1.7.1; adding missing files.

0.2.2
-----

Bugfix regarding the thermodynamic free energy array in WHAM

0.2.3
-----

Restructured the MBAR implementation for speed up and reduced the cython dependency to 0.15

0.3.0
=====

Restructured API
Implentation of new Reader
Implementation of new Data Converter (Forge)
new implementation of a runscript.
Complete new implementation of WHAM
Removed MBAR legacy code to include pymbar in futrue release
Added pytram dependency with implementation of DTRAM and XTRAM
Added example ipython notebook for asymatric double-well potential
f_K now retrun the same quantities for all estimators

0.3.1
-----
Fix Sign error in WHAM f_K.
Fix Not converged warning.
Added versioneering.
Fixing typos.

0.3.2
-----
Fix normailsation in higher thermodynamic states (WHAM).
Extended the test suite.
Renaming the *_me API functions to *_from_matrix.

0.4.0
-----
Fix to Code Style
Change to Python 3.6

DEVEL
-----





pyfeat/api/api.py
r"""

==========================
API for the pyfeat package
==========================

.. moduleauthor:: Antonia Mey <antonia.mey@fu-berlin.de>
.. moduleauthor:: Christoph Wehmeyer <christoph.wehmeyer@fu-berlin.de>

"""

from ..estimator import WHAM, XTRAM, DTRAM, TRAM
from pytram import NotConvergedWarning, ExpressionError
from ..reader import Reader
from ..forge import Forge


########################################################################
#                                                                      #
#   WHAM API function using the mathematical expressions at input      #
#                                                                      #
########################################################################


def wham_from_matrix(N_K_i, b_K_i, maxiter=100, ftol=1.0E-5, verbose=False):
    r"""
    The WHAM API function

    Parameters
    ----------
    N_K_i : numpy.ndarray(shape=(T, M), dtype=numpy.float64)
        total number of counts from simulation at T in M discrete Markov 
        state (bin)
    b_K_i : numpy.ndarray(shape=(T, M), dtype=numpy.float64)
        reduced bias energies at the T thermodynamic and M discrete Markov 
        states
    maxiter : int
        maximum number of self-consistant iteration steps during the 
        optimisation
        of the stationary probabilities
    ftol : float (> 0.0)
        convergence criterion based on the max change in an 
        self-consistent-iteration step
    verbose : boolean
        writes convergence information to stdout during the 
        self-consistent-iteration cycle

    Returns
    -------
    wham_obj : object
        WHAM estimator object with optimised unbiased stationary probabilities
    """
    # try to create the WHAM object
    try:
        wham_obj = WHAM(N_K_i, b_K_i)
    except ExpressionError as e:
        print("# ERROR ########################################################")
        print("# Your input was faulty!")
        print("# The < %s > object is malformed: %s" % (e.expression, e.msg))
        print("# ABORTING #####################################################")
        raise
    # try to converge the stationary probabilities
    try:
        wham_obj.sc_iteration(maxiter=maxiter, ftol=ftol, verbose=verbose)
    except NotConvergedWarning as e:
        print("# WARNING ######################################################")
        print("# WHAM did not converge within %d steps!" % maxiter)
        print("# The last increment was %.6e." % e.increment)
        print("# You should run the < sc_iteration > method again.")
        print("# USE RESULTS WITH CARE ########################################")
    finally:
        return wham_obj


def wham(forge, maxiter=100, ftol=1.0E-5, verbose=False):
    r"""
    The WHAM API function

    Parameters
    ----------
    forge : object
        data forge or container for pyfeat input data
    maxiter : int
        maximum number of SC iteration steps during the optimisation of the 
        stationary probabilities
    ftol : float (> 0.0)
        convergence criterion based on the max change in an 
        self-consistent-iteration step
    verbose : boolean
        writes convergence information to stdout during the 
        self-consistent-iteration cycle

    Returns
    -------
    wham_obj : object
        WHAM estimator object with optimised unbiased stationary probabilities
    """
    return wham_from_matrix(forge.N_K_i, 
                            forge.b_K_i, 
                            maxiter=maxiter, 
                            ftol=ftol, 
                            verbose=verbose)

########################################################################
#                                                                      #
#   dTRAM API function using the mathematical expressions at input     #
#                                                                      #
########################################################################


def dtram_from_matrix(C_K_ij, b_K_i, maxiter=100, ftol=1.0e-5, verbose=False):
    r"""
    Parameters
    ----------
    C_K_ij : numpy.ndarray(shape=(T, M, M), dtype=numpy.intc)
        transition counts between the M discrete Markov states for each of
        the T thermodynamic ensembles
    b_K_i : numpy.ndarray(shape=(T, M), dtype=numpy.float64)
        bias energies in the T thermodynamic and M discrete Markov states

    Returns:
    -------
    dtram_obj : obj
        dTRAM estimator object with optimised stationary properties

    """

    # try to create the DTRAM object
    try:
        dtram_obj = DTRAM(C_K_ij, b_K_i)
    except ExpressionError as e:
        print("# ERROR ########################################################")
        print("# Your input was faulty!")
        print("# The < %s > object is malformed: %s" % (e.expression, e.msg))
        print("# ABORTING #####################################################")
        raise
    # try to converge the stationary probabilities
    try:
        dtram_obj.sc_iteration(maxiter=maxiter, ftol=ftol, verbose=verbose)
    except NotConvergedWarning as e:
        print("# WARNING ######################################################")
        print("# dTRAM did not converge within %d steps!" % maxiter)
        print("# The last increment was %.6e." % e.increment)
        print("# You should run the < sc_iteration > method again.")
        print("# USE RESULTS WITH CARE ########################################")
    return dtram_obj


########################################################################
#                                                                      #
#   dTRAM API function                                                 #
#                                                                      #
########################################################################

def dtram(forge, lag=1,  maxiter=100, ftol=1.0e-5, verbose=False):
    r"""
    Parameters
    ----------
    forge : object
        data forge or container for pyfeat input data
    lag : int
        lagtime at which the countmatrix is estimated
    maxiter : int
        maximum number of SC iteration steps during the optimisation of the 
        stationary probabilities
    ftol : float (> 0.0)
        convergence criterion based on the max change in an 
        self-consistent-iteration step
    verbose : boolean
        writes convergence information to stdout during the 
        self-consistent-iteration cycle

    Returns
    -------
    dtram_obj : object
        dTRAM estimator object with optimised stationary properties
    """
    return dtram_from_matrix(forge.get_C_K_ij(lag), 
                             forge.b_K_i, 
                             maxiter=maxiter, 
                             ftol=ftol, 
                             verbose=verbose)
    
########################################################################
#                                                                      #
#   MBAR API function using the mathematical expressions at input      #
#                                                                      #
########################################################################
#def mbar_me( b_IK_x, M_I_x, N_K, maxiter=1000, ftol=1.0e-10, verbose=False ):
#    r"""
#    Parameters
#    ----------
#    b_IK_x : numpy.ndarray( shape=(T,M,M), dtype=numpy.intc )
#        transition counts between the M discrete Markov states for each of the T thermodynamic ensembles
#    M_I_x : numpy.ndarray( shape=(T,M), dtype=numpy.float64 )
#        bias energies in the T thermodynamic and M discrete Markov states
#    N_K : numpy.ndarray( shape=(T), dtype=numpy.float64 )
#        number of samples at each thermodynamic state T
#    maxiter : int
#        maximum number of SC iteration steps during the optimisation of the stationary probabilities
#    ftol : float (> 0.0)
#        convergence criterion based on the max relative change in an self-consistent-iteration step
#    verbose : boolean
#        writes convergence information to stdout during the self-consistent-iteration cycle
#    Returns:
#    -------
#    mbar_obj : obj
#        mbar estimator object with optimised stationary properties
#    """
#    # try to create the MBAR object
#    try:
#        mbar_obj = MBAR( b_IK_x, M_I_x, N_K )
#    except ExpressionError, e:
#        print "# ERROR ############################################################################"
#        print "# Your input was faulty!"
#        print "# The < %s > object is malformed: %s" % ( e.expression, e.msg )
#        print "# ABORTING #########################################################################"
#        raise
#    # try to converge the stationary probabilities
#    try:
#        mbar_obj.sc_iteration( maxiter=maxiter, ftol=ftol, verbose=verbose )
#    except NotConvergedWarning, e:
#        print "# WARNING ##########################################################################"
#        print "# WHAM did not converge within %d steps!" % maxiter
#        print "# The last relative increment was %.6e." % e.relative_increment
#        print "# You should run the < sc_iteration > method again."
#        print "# USE RESULTS WITH CARE ############################################################"
#    finally:
#        return mbar_obj


########################################################################
#                                                                      #
#   MBAR API function                                                  #
#                                                                      #
########################################################################
#def mbar( forge, maxiter = 1000, ftol = 1.0e-10, verbose = False ):
#    r"""
#    Parameters
#    ----------
#    forge : object
#        data forge or container for pyfeat input data
#    maxiter : int
#        maximum number of SC iteration steps during the optimisation of the stationary probabilities
#    ftol : float (> 0.0)
#        convergence criterion based on the max relative change in an self-consistent-iteration step
#    verbose : boolean
#        writes convergence information to stdout during the self-consistent-iteration cycle
#    Returns
#    -------
#    mbar_obj : object
#        mbar estimator object with optimised stationary properties
#    """
#    return mbar_me( forge.b_IK_x, forge.M_K_x, forge.N_K, maxiter=maxiter, ftol=ftol, verbose=verbose )

 
########################################################################
#                                                                      #
#   xTRAM API function using the mathematical expressions as input     #
#                                                                      #
########################################################################

def xtram_from_matrix(C_K_ij,
                      b_K_x, T_x, 
                      M_x, N_K_i, 
                      maxiter=100, 
                      ftol=1.0e-5, 
                      verbose=False):
    r"""
    Parameters
    ----------
    C_K_ij : numpy.ndarray(shape=(T, M, M), dtype=numpy.intc)
        transition counts between the M discrete Markov states for each of
        the T thermodynamic ensembles
    b_K_x : numpy.ndarray(shape=(T, N), dtype=numpy.float64)
        Biasing tensor
    T_x : numpy.ndarray(shape=(N,), dtype=numpy.intc)
        Thermodynamic state trajectory
    M_x : numpy.ndarray(shape=(N,), dtype=numpy.intc)
        Markov state trajectories
    N_K_i : numpy.ndarray(shape=(T, M), dtype=numpy.intc)
        Number of markov samples in each thermodynamic state

    Returns
    -------
    xtram_obj : object
        xTRAM estimator object with optimised stationary properties
    """

    # try to create the XTRAM object
    try:
        xtram_obj = XTRAM(C_K_ij, b_K_x, T_x, M_x, N_K_i)
    except ExpressionError as e:
        print("# ERROR ########################################################")
        print("# Your input was faulty!")
        print("# The < %s > object is malformed: %s" % (e.expression, e.msg))
        print("# ABORTING #####################################################")
        raise
    # try to converge the stationary probabilities
    try:
        xtram_obj.sc_iteration(maxiter=maxiter, ftol=ftol, verbose=verbose)
    except NotConvergedWarning as e:
        print("# WARNING ######################################################")
        print("# xTRAM did not converge within %d steps!" % maxiter)
        print("# The last increment was %.6e." % e.increment)
        print("# You should run the < sc_iteration > method again.")
        print("# USE RESULTS WITH CARE ########################################")
    finally:
        return xtram_obj


########################################################################
#                                                                      #
#   xTRAM API function                                                 #
#                                                                      #
########################################################################

def xtram(forge, lag=1, maxiter=100, ftol=1.0e-5, verbose=False):
    r"""
    Parameters
    ----------
    forge : object
        data forge or container for pyfeat input data
    lag : int
        lagtime at which the countmatrix is estimated
    maxiter : int
        maximum number of SC iteration steps during the optimisation of the 
        stationary probabilities
    ftol : float (> 0.0)
        convergence criterion based on the max change in an 
        self-consistent-iteration step
    verbose : boolean
        writes convergence information to stdout during the 
        self-consistent-iteration cycle

    Returns
    -------
    xtram_obj : object
        xTRAM estimator object with optimised stationary properties
    """
    return xtram_from_matrix(
        forge.get_C_K_ij(lag),
        forge.b_K_x,
        forge.T_x,
        forge.M_x,
        forge.N_K_i,
        maxiter=maxiter,
        ftol=ftol,
        verbose=verbose)


########################################################################
#                                                                      #
#   TRAM API function using the mathematical expressions at input     #
#                                                                      #
########################################################################
def tram_me():
    raise NotImplementedError('tram_me function has not been implemented yet')

########################################################################
#                                                                      #
#   TRAM API function                                                  #
#                                                                      #
########################################################################


def tram():
    raise NotImplementedError('TRAM API function has not been implemented yet')


########################################################################
#                                                                      #
#   Forge API function                                                 #
#                                                                      #
########################################################################    
    
def convert_data(trajs, b_K_i=None, kT_K=None, kT_target=None):
    r"""
    API function for creating the data forge object which will then be used 
    for the estimators

    Parameters
    ----------
    trajs : list of dictionaries
        simulation trajectories
    b_K_i : numpy.ndarray(shape=(T, M), dtype=numpy.float64)
        reduced bias energies at the T thermodynamic and M discrete Markov 
        states
    kT_K : numpy.nparray(shape=(T,), dtype=numpy.float64)
        list of kT values of each thermodynamic state
    kT_target : int
        target thermodynamic state for which pi_i and f_i will be computed

    Returns
    -------
    forge : pyfeat.Forge object
        container holding all data
    """
    forge = Forge(trajs, b_K_i, kT_K, kT_target)
    return forge

########################################################################
#                                                                      #
#   Reader API function                                                #
#                                                                      #
########################################################################


def read_files(files, 
               b_K_i_file=None, 
               kT_file=None, 
               skiprows=0, 
               maxlength=None, 
               verbose=False):
    r"""
    API function for reading files 
    
    Parameters
    ----------
    files : list of strings
        file names to read
    b_K_i_file : string (optional)
        file name for discrete estimator data
    kT_file : string (optional)
        file name for kT value listing
    skiprows : int (optional)
        specify the number of skipped lines when reading the trajectory files
    maxlength : int (optional)
        maximum number of data points read from file
    verbose : boolean (optional)
        verbose output during the reading/building process

    Returns
    -------
    reader : pyfeat.Reader object
        augmented trajectory information to be passed to the forge
    """
    reader = Reader(files, b_K_i_file, kT_file, skiprows, maxlength, verbose)
    return reader 



pyfeat/forge/forge.py
r"""

==============
Data container
==============

.. moduleauthor:: Antonia Mey <antonia.mey@fu-berlin.de>
.. moduleauthor:: Christoph Wehmeyer <christoph.wehmeyer@fu-berlin.de>

"""

import pytram as pt
import numpy as np


class Forge(pt.TRAMData):
    """I am the data forge
    """

    def __init__(self,
                 trajs,
                 b_K_i=None,
                 kT_K=None,
                 kT_target=None,
                 verbose=False):
        r"""Forge that prepares data needed for all estimators

        Parameters
        ----------

        traj : list of dictionaries 
            contianing the trajectory inforamtion
        b_K_i : np.array
            Biases
        kT_K : np.array
            array of kT values
        kT_target : int
            target value at which the probability and free energy should be 
            estimated
        verbose : boolean
            prints usefull information to stdout
        """
        super(Forge, self).__init__(trajs,
                                    b_K_i=b_K_i,
                                    kT_K=kT_K,
                                    kT_target=kT_target,
                                    verbose=verbose)
        self._M_K_x = None
        self._b_IK_x = None

    @property
    def M_K_x(self):
        if self._M_K_x is None:
            if self.verbose:
                print("# Copying Markov state sequences")
            self._M_K_x = np.zeros(shape=(self.n_therm_states, self.N_K.max()),
                                   dtype=np.intc)
            t_K = np.zeros(shape=(self.n_therm_states,), dtype=np.intc)
            for traj in self.trajs:
                # SPEEDUP POSSIBLE!
                for t in range(traj['t'].shape[0]):
                    K = traj['t'][t]
                    self._M_K_x[K, t_K[K]] = traj['m'][t]
                    t_K[K] += 1
            if self.verbose:
                print("# ... done")
        return self._M_K_x

    @property
    def b_IK_x(self):
        if self._b_IK_x is None:
            if self.verbose:
                print("# Copying bias energy sequences")
            self._b_IK_x = np.zeros(shape=(self.n_therm_states,
                                           self.n_therm_states,
                                           self.N_K.max()), dtype=np.intc)
            if not self.kT_K is None:
                self.gen_b_IK_x_from_kT_K()
            else:
                self.gen_b_IK_x()
            if self.verbose:
                print("# ... done")
        return self._b_IK_x

    def gen_b_IK_x_from_kT_K(self):
        t_K = np.zeros(shape=(self.n_therm_states,), dtype=np.intc)
        for traj in self.trajs:
            # SPEEDUP POSSIBLE!
            for t in range(traj['t'].shape[0]):
                K = traj['t'][t]
                for I in range(self.n_therm_states):
                    self._b_IK_x[I, K, t_K[K]] = self.kT_K[K] * (1.0 / self.kT_K[I]
                                                                 - 1.0 / self.kT_K[self.kT_target]) * traj['b'][t]
                t_K[K] += 1

    def gen_b_IK_x(self):
        t_K = np.zeros(shape=(self.n_therm_states,), dtype=np.intc)
        for traj in self.trajs:
            if traj['u'].shape[1] == 1:
                raise pt.ExpressionError(
                    "b_IK_x",
                    "Trajectory with single energy columns detected -"
                    "use kT file and kT target"
                )
            if traj['u'].shape[1] != self.n_therm_states:
                raise pt.ExpressionError(
                    "b_IK_x",
                    "Trajectory with wrong number of energy columns "
                    "detected (%d!=%d)" \
                    % (traj['u'].shape[1], self.n_therm_states)
                )
            # SPEEDUP POSSIBLE!
            for t in range(traj['t'].shape[0]):
                K = traj['t'][t]
                for I in range(self.n_therm_states):
                    self._b_IK_x[I, K, t_K[K]] = traj['b'][I, t]
                t_K[K] += 1


pyfeat/estimator/tram.py
r"""
======================
TRAM estimator wrapper
======================

.. moduleauthor:: Christoph Wehmeyer <christoph.wehmeyer@fu-berlin.de>

"""

import pytram as pt
import numpy as np

class TRAM( object ):
    r"""
    I am the TRAM wrapper
    """
    def __init__( self, C_K_ij, b_K_x, M_x, N_K_i ):
        r"""
        Initialize the TRAM object

        Parameters
        ----------
        C_K_ij : numpy.ndarray( shape=(T,M,M), dtype=numpy.intc )
            transition counts between the M discrete Markov states for each of the T thermodynamic ensembles
        b_K_x : numpy.ndarray( shape=(T,samples), dtype=numpy.float )
            all samples for biases at thermodynamic state K
        M_x : numpy.ndarray( shape=(samples), dtype=numpy.intc )
            trajectory of Markov states sampled
        N_K_i : numpy.ndarray( shape=(T,M), dtype=numpy.intc )
           total number of counts from simulation at T in M discrete Markov state (bin)
        """
        try:
            self._tram_obj = pt.TRAM( C_K_ij, b_K_x, M_x, N_K_i )
        except AttributeError as e:
            raise NotImplementedError( "The TRAM estimator is not yet implemented in the pytram package" )

    def sc_iteration( self , maxiter=100, ftol=1.0E-5, verbose=False ):
        r"""
        sc_iteration function

        Parameters
        ----------
        maxiter : int
            maximum number of self-consistent-iteration steps
        ftol : float (> 0.0)
            convergence criterion based on the max relative change in an self-consistent-iteration step
        verbose : boolean
            Be loud and noisy
        """
        self._tram_obj.sc_iteration( maxiter=maxiter, ftol=ftol, verbose=verbose )
         
    @property
    def pi_i( self ):
        return self._tram_obj.pi_i
     
    @property
    def pi_K_i( self ):
        return self._tram_obj.pi_K_i
    
    @property
    def f_K( self ):
        return self._tram_obj.f_K
        
    @property
    def f_K_i( self ):
        return -np.log( self.pi_K_i )
        
    @property
    def f_i( self ):
        return -np.log( self.pi_i )

    @property
    def citation( self ):
        return self._tram_obj.citation

    def cite( self, pre="" ):
        self._tram_obj.cite( pre=pre )


pyfeat/estimator/wham.py
r"""

=====================
WHAM estimator module
=====================

.. moduleauthor:: Antonia Mey <antonia.mey@fu-berlin.de>
.. moduleauthor:: Christoph Wehmeyer <christoph.wehmeyer@fu-berlin.de>

"""

import numpy as np
from pytram import NotConvergedWarning, ExpressionError


###############################################################################
#                                                                             #
#   WHAM ESTIMATOR CLASS                                                      #
#                                                                             #
###############################################################################
class WHAM(object):
    r"""
    I run the WHAM estimator
    """

    def __init__(self, N_K_i, b_K_i):
        r"""
        Initialize the WHAM object
        
        Parameters
        ----------
        N_K_i : numpy.ndarray(shape=(T, M), dtype=numpy.intc)
            total number of counts from simulation at T in M discrete Markov 
            state (bin)
        b_K_i : numpy.ndarray(shape=(T, M), dtype=numpy.float64)
            bias energies in the T thermodynamic and M discrete Markov states
        """
        self._n_therm_states = N_K_i.shape[0]
        self._n_markov_states = N_K_i.shape[1]
        self.gamma_K_i = b_K_i
        self._N_K_i = N_K_i
        self._pi_i = np.zeros(shape=(self.n_markov_states,), dtype=np.float64)
        self._f_K = None
        self._pi_K_i = None
        self.citation = [
            "The weighted histogram analysis method for free-energy"
            "calculations on biomolecules;",
            "Shankar Kumar, John M. Rosenberg, Djamal Bouzida, Robert H."
            " Swendsen and Peter A. Kollman",
            "J. Comput. Chem. 13, 1011-1021 (1992)"]

    def cite(self, pre=""):
        for line in self.citation:
            print("%s%s" % (pre, line))

    def sc_iteration(self, maxiter=100, ftol=1.0E-5, verbose=False):
        r"""
        sc_iteration function

        Parameters
        ----------
        maxiter : int
            maximum number of self-consistent-iteration steps
        ftol : float (> 0.0)
            convergence criterion based on the max relative change in a
            self-consistent-iteration step
        verbose : boolean
            Be loud and noisy
        """
        # reset internal storage variables
        self._pi_K_i = None
        self._f_K = np.ones(shape=(self.n_therm_states,), dtype=np.float64)
        finc = None
        if verbose:
            print("# %25s %25s" % ("[iteration step]", "[relative increment]"))
        if np.all(self.pi_i == 0.0):
            self._pi_i[:] = self._p_step()
        for i in range(maxiter):
            f_new = self._f_step()
            nonzero = self._f_K.nonzero()
            finc = np.max(np.abs((f_new[nonzero] -
                                  self._f_K[nonzero]) / self._f_K[nonzero]))
            self._f_K[:] = f_new[:]
            self._pi_i = self._p_step()
            if verbose:
                print("%25d %25.12e" % (i + 1, finc))
            if finc < ftol:
                break
        if finc >= ftol:
            raise NotConvergedWarning("WHAM", finc)
        self._pi_i /= self._pi_i.sum()

    def _f_step(self):
        return 1.0 / np.dot(self.gamma_K_i, self.pi_i) + np.log(self.pi_i.sum())

    def _p_step(self):
        return self._N_K_i.sum(axis=0) / np.dot(
            self._N_K_i.sum(axis=1) * self._f_K, self.gamma_K_i)

    @property
    def n_therm_states(self):
        return self._n_therm_states

    @property
    def n_markov_states(self):
        return self._n_markov_states

    @property
    def pi_i(self):
        return self._pi_i

    @property
    def f_i(self):
        return -np.log(self._pi_i)

    @property
    def f_K_i(self):
        return -np.log(self.pi_K_i)

    @property
    def f_K(self):
        if self._f_K is None:
            self._f_K = 1.0 / np.dot(self.gamma_K_i, self.pi_i)
        return np.log(self._f_K)

    @property
    def pi_K_i(self):
        if self._pi_K_i is None:
            self._pi_K_i = np.exp(self.f_K)[:, np.newaxis] * self.pi_i[np.newaxis, :] * self.gamma_K_i
        return self._pi_K_i

    ###########################################################################
    #                                                                         #
    #   _b_K_i sanity checks                                                  #
    #                                                                         #
    ###########################################################################

    def _check_b_K_i(self, b_K_i):
        if b_K_i is None:
            raise ExpressionError("b_K_i", "is None")
        if not isinstance(b_K_i, np.ndarray):
            raise ExpressionError("b_K_i", "invalid type (%s)"
                                  % str(type(b_K_i)))
        if 2 != b_K_i.ndim:
            raise ExpressionError("b_K_i", "invalid number of dimensions (%d)"
                                  % b_K_i.ndim)
        if b_K_i.shape[0] != self.n_therm_states:
            raise ExpressionError(
                "b_K_i", "unmatching number of thermodynamic states (%d,%d)"
                         % (b_K_i.shape[0], self.n_therm_states))
        if b_K_i.shape[1] != self.n_markov_states:
            raise ExpressionError(
                "b_K_i",
                "unmatching number of markov states (%d,%d)"
                % (b_K_i.shape[1], self.n_markov_states))
        if np.float64 != b_K_i.dtype:
            raise ExpressionError("b_K_i", "invalid dtype (%s)" \
                                  % str(b_K_i.dtype))
        return True

    ###########################################################################
    #                                                                         #
    #   gamma_K_i getter and setter                                           #
    #                                                                         #
    ###########################################################################

    @property
    def gamma_K_i(self):
        return self._gamma_K_i

    @gamma_K_i.setter
    def gamma_K_i(self, b_K_i):
        self._gamma_K_i = None
        if self._check_b_K_i(b_K_i):
            self._gamma_K_i = np.exp(b_K_i.min() - b_K_i)
